adding text fragments docs (#22905)
mdn/content: files/en-us/web/text_fragments/index.md

---
title: Text fragments
slug: Web/Text_fragments
tags:
  - Components
  - HTML Imports
  - JavaScript
  - Landing
  - Overview
  - Template
  - Web Article
  - Web Components
  - Web Development
  - custom elements
  - shadow dom
  - slot
browser-compat:
  - html.elements.a.text_fragments
  - api.FragmentDirective
  - css.selectors.target-text
---

> Text Fragments allow you to link directly to a specific portion of text in a web document, without needing to annotate it with an ID, using particular syntax in the URL fragment. Supporting browsers are free to choose how to draw attention to the linked text, e.g. with a color highlight and/or scrolling to the content on the page. This is useful because it allows web content authors to deep-link to specific content without relying on the presence of IDs to make that possible. Building on top of that, it could be used to generate more effective content-sharing links for users to pass to one another.

Reviewer: ```suggestion
Text Fragments allow you to link directly to a specific portion of text in a web document, without it needing to be annotated with an ID by the author, but instead using particular syntax in the URL fragment. Supporting browsers are free to choose how to draw attention to the linked text, e.g. with a color highlight and/or scrolling to the content on the page. This is useful because it allows web content consumers to deep-link to specific content without relying on the presence of web-producer-generated IDs to make that possible. Building on top of that, it could be used to generate more effective content-sharing links for users to pass to one another.
```

If you're the author, it's a lot better to generate IDs to whatever you want to link to. The actual advantage is for consumers of someone else's content, since they are not in control, but can nevertheless still place links.

Author: I've updated this so that it mitigates your concerns. I agree mostly, but I think "web-producer-generated" is a bit clumsy, and I think of consumers as being normal users who read content - they wouldn't create a link, they would just click one on a page they were reading. I think really this is more about authors linking to the content of other authors, which they don't control.

See what you think of my new wording anyhow, in the next commit.

## Concepts and usage

Historically, one of the web's key features has always been its ability to provide links between different documents — it is what makes _the web_, a web:

- You can link to the top of a document by linking to its URL, for example:
  - [https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a](/en-US/docs/Web/HTML/Element/a).

> - You can link to a specific section of a document by linking to its URL plus the _document fragment_ of that section, for example:

Reviewer: ```suggestion
- You can link to a specific section of a document by linking to its URL plus the _document fragment_ (ID) of that section, for example:
```

Author: OK, added.

The issue with linking to specific document fragments is that the author of the linked page needs to put an anchor in place to _actually_ link to. The second example above links to an {{htmlelement("h2")}} element with an ID of `browser_compatibility`:

```html
<h2 id="browser_compatibility">
  <a href="#browser_compatibility">Browser compatibility</a>
</h2>
```

If the ID is changed or removed, the document fragment is ignored, and the link just links through to the top of the page. This is reasonable in terms of graceful degradation, but it would arguably be better if the author of the link had full control over where they link to, without needing to rely on the page author.

**Text fragments** make this a reality — they allow link authors to specify text content to link to, rather than document fragments, in a flexible manner.

## Syntax

In a similar manner to document fragments, text fragments are appended onto a URL after a hash symbol (`#`). The syntax however is a bit different:

```
https://example.com#:~:text=[prefix-,]textStart[,textEnd][,-suffix]
```

The key parts to understand are as follows:

- `:~:`

>   - : Otherwise known as _the fragment directive_, this sequence of characters tells the browser that what comes next is one or more user-agent instructions, which are stripped from the URL during loading so that author scripts cannot directly interact with them. User-agent instructions are also called directives.

Reviewer: ```suggestion
  - : Otherwise known as _the fragment directive_, this sequence of characters tells the browser that what comes next are one or more user-agent instructions, which are stripped from the URL during loading so that author scripts cannot directly interact with them. User-agent instructions are also called directives.
```

Author: I think "is" reads a little bit more naturally here. You are talking about a singular "thing that comes next", which can contain one or more UA instructions.

## Examples

### Simple text fragment with textStart

- [https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#:~:text=human](/en-US/docs/Web/HTML/Element/a#:~:text=human) scrolls to and highlights the first instance of the text `human` in the document.
- [https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#:~:text=linked%20URL](/en-US/docs/Web/HTML/Element/a#:~:text=linked%20URL) scrolls to and highlights the first instance of the text `linked URL` in the document.

### textStart and textEnd

- [https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#:~:text=human,URL](/en-US/docs/Web/HTML/Element/a#:~:text=human,URL) scrolls to and highlights the first instance of a text string starting with `human` and ending with `URL`.
- [https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a#:~:text=linked%20URL,defining%20a%20value](/en-US/docs/Web/HTML/Element/a#:~:text=linked%20URL,defining%20a%20value) scrolls to and highlights the first instance of a text string starting with `linked URL` and ending with `defining a value`. Note how the highlighted text spans across multiple block-level elements.

> ### Examples with prefix- and/or suffix

Reviewer: ```suggestion
### Examples with prefix- and/or -suffix
```

Author: Good call; added

### Styling matched text fragments

Browsers are free to style the highlighted text in whatever default way they choose. The [CSS Pseudo-Elements Module Level 4](https://w3c.github.io/csswg-drafts/css-pseudo/#selectordef-target-text) defines a pseudo-element, {{cssxref("::target-text")}}, which allows you to specifying custom styling.

For example, in our [scroll-to-text demo](https://mdn.github.io/css-examples/target-text/index.html#:~:text=From%20the%20foregoing%20remarks%20we%20may%20gather%20an%20idea%20of%20the%20importance) we have the following CSS:

```css
::target-text {
  background-color: rebeccapurple;
  color: white;
}
```

Try following the above link in a supporting browser to see the effect this has.

### Programmatic access to text fragments

In supporting browsers, information on the text fragments matched in the current document can be found in the {{domxref("FragmentDirective")}} object, which is accessed via the {{domxref("Document.fragmentDirective")}} property.

> For example try running the following in Chrome devtools, in a tab with matched text fragments:

Reviewer: ```suggestion
For example try running the following in Chrome DevTools, in a tab with matched text fragments:
```

Author: See comment above
